coderd/parameters.go

```diff
@@ -188,8 +168,195 @@ func (api *API) templateVersionDynamicParameters(rw http.ResponseWriter, r *http
 	}
 }
 
-func (api *API) getWorkspaceOwnerData(
+type previewFunction func(ctx context.Context, values map[string]string) (*preview.Output, hcl.Diagnostics)
+
+func prepareDynamicPreview(ctx context.Context, rw http.ResponseWriter, db database.Store, fc *files.Cache, tf database.TemplateVersionTerraformValue, templateVersion database.TemplateVersion, user database.User) (render previewFunction, closer func(), success bool) {
+	// keep track of all files opened
+	openFiles := make([]uuid.UUID, 0)
+	closeFiles := func() {
+		for _, it := range openFiles {
+			fc.Release(it)
+		}
+	}
+
+	// This defer will close the files if the function exits early without success.
+	// Closing the files is important to avoid having a memory leak.
+	defer func() {
+		if !success {
+			closeFiles()
+		}
+	}()
+
+	// nolint:gocritic // We need to fetch the templates files for the Terraform
+	// evaluator, and the user likely does not have permission.
+	fileCtx := dbauthz.AsProvisionerd(ctx)
+	fileID, err := db.GetFileIDByTemplateVersionID(fileCtx, templateVersion.ID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error finding template version Terraform.",
+			Detail:  err.Error(),
+		})
+		return nil, nil, false
+	}
+
+	// Add the file first. Calling `Release` if it fails is a no-op, so this is safe.
+	openFiles = append(openFiles, fileID)
+	templateFS, err := fc.Acquire(fileCtx, fileID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
+			Message: "Internal error fetching template version Terraform.",
+			Detail:  err.Error(),
+		})
+		return nil, nil, false
+	}
+
+	// Having the Terraform plan available for the evaluation engine is helpful
+	// for populating values from data blocks, but isn't strictly required. If
+	// we don't have a cached plan available, we just use an empty one instead.
+	plan := json.RawMessage("{}")
+	if len(tf.CachedPlan) > 0 {
+		plan = tf.CachedPlan
+	}
+
+	openFiles = append(openFiles, tf.CachedModuleFiles.UUID)
+	if tf.CachedModuleFiles.Valid {
+		moduleFilesFS, err := fc.Acquire(fileCtx, tf.CachedModuleFiles.UUID)
+		if err != nil {
+			httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
+				Message: "Internal error fetching Terraform modules.",
+				Detail:  err.Error(),
+			})
+			return nil, nil, false
+		}
```

Comment:
this isn't even correct. this would leak the first file if the second file fails to load.

Reply:
If `sucess` is false, the files close in the defer function.

I agree the behavior is not straight forward, but it does function correctly.